src/core/server/saved_objects/mappings/lib/types.ts

```diff
@@ -0,0 +1,33 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+export interface EsMapping {
+  type?: string;
+  dynamic?: true | false | 'strict';
+  properties?: {
+    [key: string]: EsMapping;
+  };
+  fields?: {
+    [key: string]: EsMapping;
+  };
+}
+
+export interface EsMappings {
+  [rootType: string]: EsMapping;
+}
```

Comment:
So, I used to have a third type to express each property-mapping, but that's basically the same as the EsMapping. `EsMappings` unfortunately is confusing, because it really should only have a single top-level root key, which is the es mapping type. We could restrict that here in the type. Is it worth doing? One benefit is we don't really need a check for a single top level type in the IndexMappings constructor, but since this could be used without TypeScript, we may still need that check, right?

Reply:
I'm pretty sure that the root type will always be `doc`, for ever, so maybe we can encode that into our types. I'm not 100% sure about this, but if it's right it would be nice to have:

```ts
interface EsObjectPropertyMappings {
  [key: string]: EsMapping;
}

interface EsMappings {
  doc: EsObjectPropertyMappings
}
```

Reply:
It is going to be enforced as `_doc_`. At least there has been talk about that.

Reply:
Here's the reference: [dev issue](https://github.com/elastic/dev/issues/949)
and [kibana blocker for 7.0](https://github.com/elastic/kibana/issues/17568)